### ISSUE TITLE
Formalize route matching algorithm

### ISSUE THREAD

**Issue by trezm** [trezm/Thruster#75]
Given our tree structure, which I believe is fairly comprehensive, as it hasn't changed for a long time, I'd like to better formalize the algorithm for matching routes. Right now the matching code has been taped up many times and is smelling pretty bad. With a more formal algorithm we can drastically clean it up.

**Comment by ami44** [comment]
ok, but not "urgent"

**Comment by trezm** [comment]
This is an issue that should be done before the 0.5 release. It's caused many headaches and we're at a point where the code has become complex enough that it's not immediately readable by me, let alone a new contributor to the project.

**Comment by trezm** [comment]
Clearly not a priority, closing until this becomes an issue again now that the algorithm has stabilized.
